generic_config_updater/generic_updater.py

```diff
@@ -0,0 +1,469 @@
+        self.patch_applier = patch_applier
+        self.config_wrapper = config_wrapper
+        self.patch_wrapper = patch_wrapper
+
+    def replace(self, target_config):
+        if not(self.config_wrapper.validate_sonic_yang_config(target_config)):
+            raise Exception(f"The given target config is not valid")
+
+        old_config = self.config_wrapper.get_sonic_yang_as_json()
+        patch = self.patch_wrapper.generate_patch(old_config, target_config)
+
+        self.patch_applier.apply(patch)
+
+        new_config = self.config_wrapper.get_sonic_yang_as_json()
+        if not(self.patch_wrapper.verify_same_json(target_config, new_config)):
+            raise Exception(f"After applying patch to config, there is still some parts not updated")
+
+class FileSystemConfigRollbacker:
+    def __init__( \
+            self, \
+                checkpoints_dir = CHECKPOINTS_DIR, \
+                    config_replacer = ConfigReplacer(), \
+                        config_wrapper = ConfigWrapper()):
+        self.checkpoints_dir = checkpoints_dir
+        self.config_replacer = config_replacer
+        self.config_wrapper = config_wrapper
+
+    def rollback(self, checkpoint_name):
+        if not self.__check_checkpoint_exists(checkpoint_name):
+            raise Exception(f"Checkpoint '{checkpoint_name}' does not exist")
+
+        target_config = self.__get_checkpoint_content(checkpoint_name)
+
+        self.config_replacer.replace(target_config)
+
+    def checkpoint(self, checkpoint_name):
+        json_content = self.config_wrapper.get_sonic_yang_as_json()
+
+        path = self.__get_checkpoint_full_path(checkpoint_name)
+
+        self.__ensure_checkpoints_dir_exists()
+
+        self.__save_json_file(path, json_content)
+
+    def list_checkpoints(self):
+        if not self.__checkpoints_dir_exist():
+            return []
+
+        return self.__get_checkpoint_names()
+
+    def delete_checkpoint(self, checkpoint_name):
+        if not self.__check_checkpoint_exists(checkpoint_name):
+            raise Exception("Checkpoint does not exist")
+
+        self.__delete_checkpoint(checkpoint_name)
+
+    def __ensure_checkpoints_dir_exists(self):
+        os.makedirs(self.checkpoints_dir, exist_ok=True)
+
+    def __save_json_file(self, path, json_content):
+        with open(path, "w") as fh:
+            fh.write(json.dumps(json_content))
+
+    def __get_checkpoint_content(self, checkpoint_name):
+        path = self.__get_checkpoint_full_path(checkpoint_name)
+        with open(path) as fh:
+            text = fh.read()
+            return json.loads(text)
+
+    def __get_checkpoint_full_path(self, name):
+        return os.path.join(self.checkpoints_dir, f"{name}{CHECKPOINT_EXT}")
+
+    def __get_checkpoint_names(self):
+        return [f[:-len(CHECKPOINT_EXT)] for f in os.listdir(self.checkpoints_dir) if f.endswith(CHECKPOINT_EXT)]
```

Comment:
>[f[:-len(CHECKPOINT_EXT)] for f in os.listdir(self.checkpoints_dir) if f.endswith(CHECKPOINT_EXT)] [](http://example.com/codeflow?start=15&length=98)

This line is complex. Better to use a normal loop and add a block comment to explain your method, even better with an example.